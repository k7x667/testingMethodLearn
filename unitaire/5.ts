/**
 * Vous avez ici un store zustand qui vous permet de gérer des infos d'utilisateurs ainsi que quelques photos
 * vérifiez que les fonctions affectent bien le store comme prévu
 *
 * Attention pensez bien à reset le store avant chaque tests (il se peut qu'il y ai quelque chose sur ça pour vitest ou jest)
 *
 */

import create from 'zustand';

export interface User {
  id: number;
  name: string;
}

export interface Picture {
  id: number;
  userId: number;
  url: string;
}

export interface UserStore {
  users: User[];
  currentUser: User | null;
  pictures: Picture[];
  setCurrentUser: (user: User | null) => void;
  addPicture: (picture: Picture) => void;
  deletePicture: (pictureId: number) => void;
  reset: () => void;
}

export const useUserStore = create<UserStore>((set) => ({
  users: [],
  currentUser: null,
  pictures: [],
  setCurrentUser: (user) => set((state) => ({ ...state, currentUser: user })),
  addPicture: (picture) =>
    set((state) => ({ ...state, pictures: [...state.pictures, picture] })),
  deletePicture: (pictureId) =>
    set((state) => ({
      ...state,
      pictures: state.pictures.filter((picture) => picture.id !== pictureId),
    })),
  reset: () => set(() => ({ users: [], currentUser: null, pictures: [] })),
}));
