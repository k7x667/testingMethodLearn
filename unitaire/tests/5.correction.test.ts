import { test, expect, vi, describe, beforeEach } from 'vitest';
import { useUserStore } from '../5';

describe('Store Zustand user', () => {
  beforeEach(() => {
    const { reset } = useUserStore.getState();
    reset();
  });

  test('Test can add a picture in the picture array', () => {
    const { addPicture } = useUserStore.getState();

    addPicture({
      id: 1,
      userId: 1,
      url: 'https://google.com',
    });

    const { pictures } = useUserStore.getState();
    expect(pictures).toMatchSnapshot();
  });

  test('Test can add multiple element in the picture array', () => {
    const { addPicture } = useUserStore.getState();

    addPicture({
      id: 1,
      userId: 1,
      url: 'https://google.com',
    });
    addPicture({
      id: 2,
      userId: 2,
      url: 'https://google.com',
    });

    const { pictures } = useUserStore.getState();
    expect(pictures).toMatchSnapshot();
  });

  test('Test can delete a picture in the picture array', () => {
    const { addPicture, deletePicture } = useUserStore.getState();

    addPicture({
      id: 1,
      userId: 1,
      url: 'https://google.com',
    });
    deletePicture(1);

    const { pictures } = useUserStore.getState();
    expect(pictures).toEqual([]);
  });

  test('Test can add a picture in mutiple picture array', () => {
    const { addPicture, deletePicture } = useUserStore.getState();

    addPicture({
      id: 1,
      userId: 1,
      url: 'https://google.com',
    });
    addPicture({
      id: 2,
      userId: 2,
      url: 'https://google.com',
    });
    deletePicture(1);

    const { pictures } = useUserStore.getState();
    expect(pictures).toMatchSnapshot();
  });

  test('Test can add a picture in mutiple picture array', () => {
    const { deletePicture } = useUserStore.getState();
    deletePicture(1);

    const { pictures } = useUserStore.getState();
    expect(pictures).toEqual([]);
  });

  test('Test can add a current user', () => {
    const { setCurrentUser } = useUserStore.getState();
    setCurrentUser({
      id: 1,
      name: 'test',
    });

    const { currentUser } = useUserStore.getState();
    expect(currentUser).toEqual({
      id: 1,
      name: 'test',
    });
  });

  test('Test can reset the data inside the store', () => {
    const { setCurrentUser, addPicture, reset } = useUserStore.getState();
    setCurrentUser({
      id: 1,
      name: 'test',
    });
    addPicture({
      id: 1,
      userId: 1,
      url: 'https://google.com',
    });

    reset();

    const { currentUser, pictures, users } = useUserStore.getState();

    expect(currentUser).toBeNull();
    expect(pictures).toEqual([]);
    expect(users).toEqual([]);
  });
});
