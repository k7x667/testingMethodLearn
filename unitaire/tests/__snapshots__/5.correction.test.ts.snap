// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Store Zustand user > Test can add a picture in mutiple picture array 1`] = `
[
  {
    "id": 2,
    "url": "https://google.com",
    "userId": 2,
  },
]
`;

exports[`Store Zustand user > Test can add a picture in the picture array 1`] = `
[
  {
    "id": 1,
    "url": "https://google.com",
    "userId": 1,
  },
]
`;

exports[`Store Zustand user > Test can add multiple element in the picture array 1`] = `
[
  {
    "id": 1,
    "url": "https://google.com",
    "userId": 1,
  },
  {
    "id": 2,
    "url": "https://google.com",
    "userId": 2,
  },
]
`;
